### PR TITLE
fix: Migrate all rules from eslint-plugin-twenty to eslint-plugin-twenty-ts

### LINKED ISSUE #1590: Migrate matching-state-variable rule from eslint-plugin-twenty to eslint-plugin-twenty-ts
## Context

We are maintaining a package `eslint-plugin-twenty` to add our own linter rule and help the community share the same standards. We have introduced a new `eslint-plugin-twenty-ts` package where rules can be written in typescript and tested. We need to migrate from `eslint-plugin-twenty` to `eslint-plugin-twenty-ts`

## Scope
This ticket is about moving the `matching-state-variable` rule from `eslint-plugin-twenty` to `eslint-plugin-twenty-ts`. 
- take a look at `twenty/packages/eslint-plugin-twenty/rules/matching-state-variable.js`
- create a new rule within eslint-plugin-twenty-ts that does the same thing
- write tests to cover the different cases  'useRecoilState' / 'useRecoilFamilyState' / 'useRecoilSelector' / 'useRecoilScopedState' / 'useRecoilScopedFamilyState' / 'useRecoilScopedSelector' / 'useRecoilValue'
- make sure the linting rule is still working by trying to badly name recoilState: `const currentUserBla = useRecoilValue(currentUserState);` should be an error

### LINKED ISSUE #1591: Migrate no-hardcoded-colors rule from eslint-plugin-twenty to eslint-plugin-twenty-ts
## Context

We are maintaining a package `eslint-plugin-twenty` to add our own linter rule and help the community share the same standards. We have introduced a new `eslint-plugin-twenty-ts` package where rules can be written in typescript and tested. We need to migrate from `eslint-plugin-twenty` to `eslint-plugin-twenty-ts`

## Scope
This ticket is about moving the `no-hardcoded-colors` rule from `eslint-plugin-twenty` to `eslint-plugin-twenty-ts`. 
- take a look at `twenty/packages/eslint-plugin-twenty/rules/no-hardcoded-colors.js`
- create a new rule within eslint-plugin-twenty-ts that does the same thing
- write tests to cover the different case
- make sure that the linter is still detecting hardcoded colors in the codebase

### LINKED ISSUE #1592: Migrate sort-css-properties-alphabetically rule from eslint-plugin-twenty to eslint-plugin-twenty-ts
## Context

We are maintaining a package `eslint-plugin-twenty` to add our own linter rule and help the community share the same standards. We have introduced a new `eslint-plugin-twenty-ts` package where rules can be written in typescript and tested. We need to migrate from `eslint-plugin-twenty` to `eslint-plugin-twenty-ts`

## Scope
This ticket is about moving the `no-hardcoded-colors` rule from `eslint-plugin-twenty` to `eslint-plugin-twenty-ts`. 
- take a look at `twenty/packages/eslint-plugin-twenty/rules/sort-css-properties-alphabetically.js`
- create a new rule within eslint-plugin-twenty-ts that does the same thing
- write tests to cover the different case
- make sure that the linter is still detecting non ordered imports

### LINKED ISSUE #1593: Migrate styled-components-prefixed-with-styled rule from eslint-plugin-twenty to eslint-plugin-twenty-ts
## Context

We are maintaining a package `eslint-plugin-twenty` to add our own linter rule and help the community share the same standards. We have introduced a new `eslint-plugin-twenty-ts` package where rules can be written in typescript and tested. We need to migrate from `eslint-plugin-twenty` to `eslint-plugin-twenty-ts`

## Scope
This ticket is about moving the `no-hardcoded-colors` rule from `eslint-plugin-twenty` to `eslint-plugin-twenty-ts`. 
- take a look at `twenty/packages/eslint-plugin-twenty/rules/styled-components-prefixed-with-styled.js`
- create a new rule within eslint-plugin-twenty-ts that does the same thing
- write tests to cover the different case
- make sure that the linter is still detecting styled component with a name not prefixed by Styled

### PR DESCRIPTION
**Pull request by anikdhabal**
fixes #1590
fixes #1591 
fixes #1592 
fixes #1593

### CODE REVIEW

**Review by charlesBochet — COMMENTED** [review]
Wow looks great! Can we completely remove the old linter plugin? (eslint-plugin-twenty)

**Comment by anikdhabal** [comment]
> Wow looks great! Can we completely remove the old linter plugin? (eslint-plugin-twenty)

Yes we can remove.

File: server/.eslintrc.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ module.exports = {
     tsconfigRootDir : __dirname, 
     sourceType: 'module',
   },
-  plugins: ['@typescript-eslint/eslint-plugin', 'import', 'unused-imports'],
+  plugins: ['@typescript-eslint/eslint-plugin', 'import', 'unused-imports','twenty-ts'],
```
Comment (charlesBochet) [inline server/.eslintrc.js:8]: You'll need to add the eslint-plugin-twenty-ts to server package.json and to add its installation to server yarn install command (similar to what has been done in front/package.json)
I suspect this might be not straightforward + you'll need to fix many linter errors, so let's do that in another PR :)

Comment (anikdhabal) [reply server/.eslintrc.js:8]: I am going fix those now. Thanks for your review.

File: packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
+import { Rule } from 'eslint';
```
Comment (charlesBochet) [inline packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts:1]: could you use the same format as effect-component.ts to define these rules?

Comment (charlesBochet) [reply packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts:1]: Being consistent across the codebase is one of our focus :)

File: packages/eslint-plugin-twenty-ts/src/rules/no-hardcoded-colors.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { Rule } from "eslint";
+
+const rule: Rule.RuleModule = {
```
Comment (akhill10) [inline packages/eslint-plugin-twenty-ts/src/rules/no-hardcoded-colors.ts:3]: `meta` values are missing in this rule.

File: packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts (hunk 2 above)
Comment (anikdhabal) [reply packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts:1]: OKK

File: packages/eslint-plugin-twenty-ts/src/rules/no-hardcoded-colors.ts (hunk 3 above)
Comment (anikdhabal) [reply packages/eslint-plugin-twenty-ts/src/rules/no-hardcoded-colors.ts:3]: Ahh, oversight, thank you.

File: packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts (hunk 2 above)
Comment (anikdhabal) [reply packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts:1]: @charlesBochet Done update the effect-component.ts and the test.

File: server/.eslintrc.js (hunk 1 above)
Comment (anikdhabal) [reply server/.eslintrc.js:8]: Yes, let's fix all the linters errors in another PR.

File: packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts (hunk 2 above)
Comment (yashug) [reply packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts:1]: @anikdhabal i think we need to follow the format from effect-component.ts, but not the other way around

Comment (charlesBochet) [reply packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts:1]: yes exactly :) Also we don't use export default in the project :)

File: packages/eslint-plugin-twenty-ts/src/tests/all.spec.test.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
+import { RuleTester } from "@typescript-eslint/rule-tester";
```
Comment (charlesBochet) [inline packages/eslint-plugin-twenty-ts/src/tests/all.spec.test.ts:1]: let's rename this file effect-component.spec.ts

I've also notice that test files are inconsistently named. They should all be suffixed with .spec.ts :)

Comment (anikdhabal) [reply packages/eslint-plugin-twenty-ts/src/tests/all.spec.test.ts]: done

File: packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts (hunk 2 above)
Comment (anikdhabal) [reply packages/eslint-plugin-twenty-ts/src/rules/matching-state-variable.ts:1]: fixed

**Comment by anikdhabal** [comment]
@charlesBochet can I delete all the old packages in this PR or make another PR for that along with some linter fix?

**Comment by yashug** [comment]
@anikdhabal please also consider updating this file also with new ts rules https://github.com/twentyhq/twenty/blob/main/front/.eslintrc.js

**Comment by anikdhabal** [comment]
> @anikdhabal please also consider updating this file also with new ts rules https://github.com/twentyhq/twenty/blob/main/front/.eslintrc.js

Done, thanks mate

**Comment by anikdhabal** [comment]
@charlesBochet Fixed and Updated all related problem. Pls review once. Removed the old packages, fixed lint problem, followed the format from effect-component.ts.

**Review by charlesBochet — APPROVED** [review]
Merging it

**Comment by anikdhabal** [comment]
> Merging it

Thanks a lot.

**Comment by charlesBochet** [comment]
@anikdhabal I have used code from your PR and re-applied some from neo so, I will still split the bounty.
I'm in touch with Algora so this won't happen again :)

**Comment by charlesBochet** [comment]
Regarding the PR, the tests of the packages are broken:

```
cd ./packages/eslint-plugin-twenty-ts 
yarn test
```

Could you take a look? If you can't that's fine, I can take care of them later :)

**Comment by anikdhabal** [comment]
> Regarding the PR, the tests of the packages are broken:
> 
> ```
> cd ./packages/eslint-plugin-twenty-ts 
> yarn test
> ```
> 
> Could you take a look? If you can't that's fine, I can take care of them later :)

Why not, let me handle it.

**Comment by anikdhabal** [comment]
> @anikdhabal I have used code from your PR and re-applied some from neo so, I will still split the bounty. I'm in touch with Algora so this won't happen again :)

Not an issue. Thank you again.

**Comment by anikdhabal** [comment]
> /tip $100 @anikdhabal

How to claim this one? I've seen that it automatically added, but here it redirects me to the installation page.

**Comment by charlesBochet** [comment]
It has been sent, thank you :)
